### PR TITLE
🌱 Validate subnet's az with failure domain when subnet id is passed

### LINKED ISSUE #1838: Specifying subnetID for an AWSMachine and a conflicting failureDomain will ignore the failureDomain
/kind bug

Summarizing from this comment: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1833#discussion_r459665277

**What steps did you take and what happened:**
- create an AWSMachine while specifying the subnetID it should attach to
- also specify a failureDomain that **isn't** the failureDomain of the subnet

**What did you expect to happen:**
Expect it to validate the subnet is in the correct failureDomain and run into errors trying to create the instance. 
Actual: machine is attached to subnet specified by subnetID

### PR DESCRIPTION
**Pull request by shysank**
**What this PR does / why we need it**:
Validates subnet's availability zone with failure domain when subnet id is passed

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1838

### CODE REVIEW

File: pkg/cloud/services/ec2/instances.go [hunk 1]
```diff
@@ -262,26 +262,54 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 
 	switch {
 	case scope.AWSMachine.Spec.Subnet != nil && scope.AWSMachine.Spec.Subnet.ID != nil:
+		if failureDomain != nil {
+			subnets, err := s.getFilteredSubnets(filter.EC2.SubnetID(aws.StringValue(scope.AWSMachine.Spec.Subnet.ID)))
```
Comment (shysank) [inline pkg/cloud/services/ec2/instances.go:266]: Should VPC ID also be passed here?

**Comment by gab-satchi** [comment]
linter giving a suggestion for one of the for loops: 
```pkg/cloud/services/ec2/instances.go:355:2: S1011: should replace loop with `subnetInput.Filters = append(subnetInput.Filters, criteria...)` (gosimple)```

File: pkg/cloud/filter/ec2.go [hunk 2]
```diff
@@ -156,3 +157,10 @@ func (ec2Filters) IgnoreLocalZones() *ec2.Filter {
 		Values: aws.StringSlice([]string{"opt-in-not-required"}),
 	}
 }
+
```
Comment (richardcase) [inline pkg/cloud/filter/ec2.go:160]: Small nit, a function comment would be a nice to have but not all the exported functions in this file have them

File: pkg/cloud/services/ec2/instances.go [hunk 3]
```diff
@@ -262,26 +262,54 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
...
+			if err != nil {
+				return "", err
```
Comment (richardcase) [inline pkg/cloud/services/ec2/instances.go:268]: Could potentially wrap the error, but a nice to have

**Review by richardcase — APPROVED** [review]
Thanks @shysank for this change. A couple of minor comments but these are only nice to haves, so:

/lgtm

File: pkg/cloud/filter/ec2.go (hunk 2 above)
Comment (shysank) [reply pkg/cloud/filter/ec2.go:160]: Fixed it.

File: pkg/cloud/services/ec2/instances.go (hunk 3 above)
Comment (shysank) [reply pkg/cloud/services/ec2/instances.go]: Fixed it, Thanks!

File: pkg/cloud/services/ec2/instances.go [hunk 4]
```diff
@@ -262,26 +262,55 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
...
+				return "", errors.Wrapf(err, "failed to filter subnets for criteria %q", criteria)
+			}
+
+			if len(subnets) == 0 {
+				record.Warnf(scope.AWSMachine, "FailedCreate",
+					"Failed to create instance: subnet with id %q not found", aws.StringValue(scope.AWSMachine.Spec.Subnet.ID))
+				return "", awserrors.NewFailedDependency(
+					fmt.Sprintf("failed to run machine %q, subnet with id %q not found",
+						scope.Name(),
+						aws.StringValue(scope.AWSMachine.Spec.Subnet.ID),
```
Comment (randomvariable) [inline pkg/cloud/services/ec2/instances.go:278]: Can we introspect the subnet information from the AWSCluster object instead? I'm wary of adding an additional EC2 API call if we can avoid it.

Comment (shysank) [reply pkg/cloud/services/ec2/instances.go:278]: Do you mean something like this `s.scope.Subnets().FindByID("id")` ?

Comment (randomvariable) [reply pkg/cloud/services/ec2/instances.go:278]: Yup. It might be we need to persist more information in AWSCluster status to enable this.

Comment (shysank) [reply pkg/cloud/services/ec2/instances.go:278]: I see the subnets being populated [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/a229052b19a7b4d976f312a6bf6a42a9ce09b5d8/pkg/cloud/services/network/subnets.go#L148), but I am not entirely sure, what extra data need to go in there. Is there some documentation that I can refer to?
Should we also modify `getFilteredSubnets` to look for subnets in status, and fallback to ec2 api if it's not there?

Comment (richardcase) [reply pkg/cloud/services/ec2/instances.go:278]: I think there should be everything needed already persisted:

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/a229052b19a7b4d976f312a6bf6a42a9ce09b5d8/pkg/cloud/services/network/subnets.go#L264

Does that have everything needed @shysank ?

Comment (shysank) [reply pkg/cloud/services/ec2/instances.go]: I've made the changes, PTAL.

**Comment by richardcase** [comment]
This is looking good to me.

@shysank - would you be able to squash your commits?

**Pull request merged**